### PR TITLE
Add `query` table function for query pass-through to Cassandra connector

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Add `query` table function for query pass-through to Cassandra connector

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Cassandra
* Add `query` table function for query pass-through to the connector. ({issue}`15973`)
```

### CODE REVIEW

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTableHandle.java [hunk 1]
```diff
@@ -14,79 +14,107 @@
...
 public class CassandraTableHandle
         implements ConnectorTableHandle
 {
-    private final String schemaName;
-    private final String tableName;
-    private final Optional<List<CassandraPartition>> partitions;
-    private final String clusteringKeyPredicates;
+    private final CassandraRelationHandle relationHandle;
 
+    @Deprecated
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTableHandle.java:35]: Instead of marking it as deprecated - can we remove the usage of this constructor. I think it is being used in 4 places. Maybe it could go as a preparatory commit or a different PR in the followup

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTableHandle.java [hunk 2]
```diff
@@ -14,79 +14,107 @@
...
+    public boolean isNamedRelation()
+    {
+        return relationHandle instanceof CassandraNamedRelationHandle;
+    }
+
+    @JsonIgnore
+    public CassandraNamedRelationHandle asPlainTable()
+    {
+        checkState(!isSynthetic(), "The table handle does not represent a plain table: %s", this);
+        return getRequiredNamedRelation();
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTableHandle.java:82]: Isn't this API a bit similar to `getRequiredNamedRelation` or can we have one copy of it.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTableHandle.java [hunk 3]
```diff
@@ -14,79 +14,107 @@
...
+        return asPlainTable().getSchemaName();
     }
 
-    @JsonProperty
+    @JsonIgnore
     public String getTableName()
     {
-        return tableName;
+        return asPlainTable().getTableName();
     }
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTableHandle.java:95]: Schema and table are not a part of `CassandraTableHandle` now - instead of having an API here - can we ask caller to get it as `asPlainTable` or `getRequiredNamedRelation` and then get the schema or tableName accordingly ?

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraMetadata.java [hunk 4]
```diff
@@ -209,7 +230,7 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
     public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(ConnectorSession session, ConnectorTableHandle tableHandle, Constraint constraint)
     {
         CassandraTableHandle handle = (CassandraTableHandle) tableHandle;
-        if (handle.getPartitions().isPresent() || !handle.getClusteringKeyPredicates().isEmpty()) {
+        if (handle.isSynthetic() || handle.getPartitions().isPresent() || !handle.getClusteringKeyPredicates().isEmpty()) {
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraMetadata.java:233]: Can we have a separate `if` statement for this ? like JdbcBased connector we can also add the filter as as additional filter criteria right ? And we could have a dedicated //TODO for the same.

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestJsonCassandraHandles.java [hunk 5]
```diff
@@ -37,20 +37,24 @@
 public class TestJsonCassandraHandles
 {
     private static final Map<String, Object> TABLE_HANDLE_AS_MAP = ImmutableMap.of(
-            "schemaName", "cassandra_schema",
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestJsonCassandraHandles.java:40]: As a follow up - can we use `assertJsonRoundTrip` for testing these handles ?

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java [hunk 6]
```diff
@@ -1384,6 +1386,71 @@ public void testRowLevelDelete()
                 .hasStackTraceContaining("Delete without primary key or partition key is not supported");
     }
 
+    // test polymorphic table function
+
+    @Test
+    public void testNativeQuerySelectFromNation()
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1392]: Can we add additional test to ensure that `filterPushDown` doesn't happen with this PTF and also one with invalid column type.

Comment (ebyhr) [reply plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1392]: Added, but the remaining unsupported type in Cassandra connector is only time type and #15802 is going to support the type.

**Comment by ebyhr** [comment]
Addressed comments.

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java (hunk 6 above)
Comment (Praveen2112) [reply plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1392]: Is there is any type in cassandra - which is difficult to support or will not be supported in future. Or maybe some user defined type.

Comment (ebyhr) [reply plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1392]: I don't there is. 25/26 types including UDT are supported.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/ptf/Query.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            this.cassandraMetadata = requireNonNull(cassandraMetadata, "metadata is null");
+        }
+
+        @Override
+        public TableFunctionAnalysis analyze(ConnectorSession session, ConnectorTransactionHandle transaction, Map<String, Argument> arguments)
+        {
+            ScalarArgument argument = (ScalarArgument) getOnlyElement(arguments.values());
+            String query = ((Slice) argument.getValue()).toStringUtf8();
+
+            CassandraQueryRelationHandle queryRelationHandle = new CassandraQueryRelationHandle(query);
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/ptf/Query.java:97]: Can we inline `CassandraQueryRelationHandle`

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java [hunk 8]
```diff
@@ -1384,6 +1387,85 @@ public void testRowLevelDelete()
...
+                .isNotFullyPushedDown(FilterNode.class);
+    }
+
+    @Test
+    public void testNativeQuerySelectFromTestTable()
+    {
+        String tableName = "tpch.test_select" + randomNameSuffix();
+        onCassandra("CREATE TABLE " + tableName + "(col BIGINT PRIMARY KEY)");
+
+        onCassandra("INSERT INTO " + tableName + "(col) VALUES (1)");
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1411]: Can we use `testTable` so we don't have to have a dedicated query executor for cassandra  and we don't have to drop the table.

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java [hunk 9]
```diff
@@ -1384,6 +1387,85 @@ public void testRowLevelDelete()
...
+    public void testNativeQueryInsertStatementTableExists()
+    {
+        String tableName = "test_insert" + randomNameSuffix();
+        onCassandra("CREATE TABLE tpch." + tableName + "(col INT PRIMARY KEY)");
+
+        assertThatThrownBy(() -> query("SELECT * FROM TABLE(cassandra.system.query(query => 'INSERT INTO tpch." + tableName + "(col) VALUES (3)'))"))
+                .hasMessage("Handle doesn't have columns info");
+
+        onCassandra("DROP TABLE IF EXISTS tpch." + tableName);
+    }
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1448]: In case of `DELETE` what is our stand ?

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java [hunk 10]
```diff
@@ -1384,6 +1387,85 @@ public void testRowLevelDelete()
...
+        assertFalse(getQueryRunner().tableExists(getSession(), tableName));
+    }
+
+    @Test
+    public void testNativeQueryInsertStatementTableDoesNotExist()
+    {
+        String tableName = "test_insert" + randomNameSuffix();
+        assertFalse(getQueryRunner().tableExists(getSession(), tableName));
+        assertThatThrownBy(() -> query("SELECT * FROM TABLE(cassandra.system.query(query => 'INSERT INTO tpch." + tableName + "(col) VALUES (1)'))"))
+                .hasMessageContaining("unconfigured table");
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1435]: What's the rationale for this test ?

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestJsonCassandraHandles.java [hunk 11]
```diff
@@ -37,20 +37,24 @@
 public class TestJsonCassandraHandles
 {
     private static final Map<String, Object> TABLE_HANDLE_AS_MAP = ImmutableMap.of(
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestJsonCassandraHandles.java:39]: We need a test for `CassandraQueryRelationHandle` right ?

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/util/CassandraCqlUtils.java [hunk 12]
```diff
@@ -116,8 +116,8 @@ public static Select selectFrom(CassandraTableHandle tableHandle, List<Cassandra
 
     public static SelectFrom from(CassandraTableHandle tableHandle)
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/util/CassandraCqlUtils.java:117]: WDYT - Instead of passing `CassandraTableHandle` can we pass `CassandraNamedRelationHandle`. 

Just an idea - It could be implemented as a follow up or a preparatory commit.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTableHandle.java (hunk 1 above)
Comment (Praveen2112) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraTableHandle.java:34]: Can the changes related to removal of the Deprecated constructor - be a different commit.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSplitManager.java [hunk 13]
```diff
@@ -125,16 +129,18 @@ private String extractClusteringKeyPredicates(CassandraPartitionResult partition
...
     }
 
     private List<ConnectorSplit> getSplitsByTokenRange(CassandraTable table, String partitionId, Optional<Long> sessionSplitsPerNode)
     {
-        String schema = table.getTableHandle().getSchemaName();
-        String tableName = table.getTableHandle().getTableName();
+        CassandraTableHandle handle = table.getTableHandle();
+        String schema = handle.getRequiredNamedRelation().getSchemaName();
+        CassandraTableHandle handle1 = table.getTableHandle();
+        String tableName = handle1.getRequiredNamedRelation().getTableName();
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSplitManager.java:143]: Why can't we use handle here ?

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSplitManager.java [hunk 14]
```diff
@@ -125,16 +129,18 @@ private String extractClusteringKeyPredicates(CassandraPartitionResult partition
...
                 partitionResult.getUnenforcedConstraint(),
                 session.getCassandraVersion());
         return clusteringPredicatesExtractor.getClusteringKeyPredicates();
     }
 
     private List<ConnectorSplit> getSplitsByTokenRange(CassandraTable table, String partitionId, Optional<Long> sessionSplitsPerNode)
     {
-        String schema = table.getTableHandle().getSchemaName();
-        String tableName = table.getTableHandle().getTableName();
+        CassandraTableHandle handle = table.getTableHandle();
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSplitManager.java:140]: Can have it as `table.getTableHandle().getRequiredNamedRelation()` instead

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java [hunk 15]
```diff
@@ -270,7 +270,7 @@ public CassandraTable getTable(SchemaTableName schemaTableName)
                 .sorted(comparing(CassandraColumnHandle::getOrdinalPosition))
                 .collect(toList());
 
-        CassandraTableHandle tableHandle = new CassandraTableHandle(tableMeta.getKeyspace().asInternal(), tableMeta.getName().asInternal());
+        CassandraTableHandle tableHandle = new CassandraTableHandle(new CassandraNamedRelationHandle(tableMeta.getKeyspace().asInternal(), tableMeta.getName().asInternal()));
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java:273]: What is your opinion on returning as `CassandraNamedRelationHandle` ?

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraRecordCursor.java [hunk 16]
```diff
@@ -24,21 +25,26 @@
 
 import java.util.List;
 
+import static com.google.common.base.Preconditions.checkArgument;
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraRecordCursor.java:28]: This change one moving from an index based reading data to a columnName based - can be moved to a different commit. WDYT ?

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestJsonCassandraHandles.java (hunk 11 above)
Comment (Praveen2112) [reply plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestJsonCassandraHandles.java:39]: It is always nice to have a roundtrip test right ?

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java (hunk 10 above)
Comment (ebyhr) [reply plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1435]: To verify failure from Cassandra, not connector side.

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java (hunk 8 above)
Comment (ebyhr) [reply plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1411]: Note that `TestCassandraTable` doesn't support single column with primary key.

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestJsonCassandraHandles.java (hunk 11 above)
Comment (ebyhr) [reply plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestJsonCassandraHandles.java:39]: `TestCassandraTableHandle` already exists. I will add another method there.

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java (hunk 9 above)
Comment (ebyhr) [reply plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1448]: Same as `INSERT` statement. I will add the case.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java (hunk 15 above)
Comment (ebyhr) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java:273]: Could you elaborate on that?

**Comment by ebyhr** [comment]
Addressed comments.

**Comment by ebyhr** [comment]
I found a correctness issue when duplicated columns exist with different cases. e.g. col & COL. I'm fixing the issue now.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraMetadata.java [hunk 17]
```diff
@@ -119,16 +125,31 @@ public CassandraTableHandle getTableHandle(ConnectorSession session, SchemaTable
...
     }
 
     @Override
     public ConnectorTableMetadata getTableMetadata(ConnectorSession session, ConnectorTableHandle tableHandle)
     {
         requireNonNull(tableHandle, "tableHandle is null");
+        CassandraTableHandle handle = (CassandraTableHandle) tableHandle;
+        if (handle.getRelationHandle() instanceof CassandraQueryRelationHandle queryRelationHandle) {
+            List<ColumnMetadata> columns = getColumnHandles(queryRelationHandle.getQuery()).stream()
+                    .map(column -> ((CassandraColumnHandle) column).getColumnMetadata())
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraMetadata.java:138]: Can we split this into two map like `map(CassandraColumnHandle.class::cast).map(CassandraColumnHandle::getColumnMetadata)

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraMetadata.java [hunk 18]
```diff
@@ -119,16 +125,31 @@ public CassandraTableHandle getTableHandle(ConnectorSession session, SchemaTable
...
+                    .collect(toImmutableList());
+            return new ConnectorTableMetadata(getSchemaTableName(handle), columns, ImmutableMap.of(), Optional.empty());
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraMetadata.java:140]: Why don't we use this ctr `ConnectorTableMetadata(getSchemaTableName(handle), columns)` ?

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraMetadata.java [hunk 19]
```diff
@@ -351,11 +377,11 @@ public ConnectorInsertTableHandle beginInsert(ConnectorSession session, Connecto
         }
 
         CassandraTableHandle table = (CassandraTableHandle) tableHandle;
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraMetadata.java:379]: If we can change the above line as `CassandraNameRelationHandle table = ((CassandraTableHandle) tableHandle).getRequiredNamedRelation()

Then most of git diff can be avoided right ?

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java (hunk 15 above)
Comment (Praveen2112) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java:273]: Currently `CassandraTable` has two fields - `CassandraTableHandle` and `List<CassandraColumnHandle`. `CassandraTableHandle` is wrapper for both named and query based relation - which might be an overkill here. Instead can we have something like this 
`public CassandraTable(CassandraNamedRelationHandle tableHandle, List<CassandraColumnHandle> columns)`

Just thinking out aloud.

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java (hunk 8 above)
Comment (Praveen2112) [reply plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1411]: In this case can we move dropping operations to `finally` block ?

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java [hunk 20]
```diff
@@ -1384,6 +1387,101 @@ public void testRowLevelDelete()
...
+    public void testNativeQueryUnsupportedStatement()
+    {
+        String tableName = "test_unsupported_statement" + randomNameSuffix();
+        onCassandra("CREATE TABLE tpch." + tableName + "(col INT PRIMARY KEY)");
+
+        assertThatThrownBy(() -> query("SELECT * FROM TABLE(cassandra.system.query(query => 'INSERT INTO tpch." + tableName + "(col) VALUES (3)'))"))
+                .hasMessage("Handle doesn't have columns info");
+        assertThatThrownBy(() -> query("SELECT * FROM TABLE(cassandra.system.query(query => 'DELETE FROM tpch." + tableName + " WHERE col = 1'))"))
+                .hasMessage("Handle doesn't have columns info");
+
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1462]: Can we add a check to ensure if table is not yet deleted ?

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java (hunk 8 above)
Comment (Praveen2112) [reply plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1411]: Since we use `onCassandra` at a some places - Can we add support for `TestCassandraTable` with a single column primary key ?

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java [hunk 21]
```diff
@@ -1384,6 +1387,101 @@ public void testRowLevelDelete()
                 .hasStackTraceContaining("Delete without primary key or partition key is not supported");
     }
 
+    // test polymorphic table function
+
+    @Test
+    public void testNativeQuerySelectFromNation()
+    {
+        assertQuery(
+                "SELECT * FROM TABLE(cassandra.system.query(query => 'SELECT name FROM tpch.nation WHERE nationkey = 0 ALLOW FILTERING'))",
```
Comment (Praveen2112) [inline plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1396]: Can we add an additional test case like `"SELECT name FROM TABLE(cassandra.system.query(query => 'SELECT * FROM tpch.nation WHERE nationkey = 0 ALLOW FILTERING'`

**Review by Praveen2112 — APPROVED** [review]
LGTM. Minor comments. Thanks for working on this.

File: plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java (hunk 8 above)
Comment (ebyhr) [reply plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestCassandraConnectorTest.java:1414]: Let me handle in follow-up PR.
